substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -758,33 +1304,50 @@ impl<T: Config> Pallet<T> {
             Error::<T>::CannotUpdateContractInGraceState
         );
 
-        let mut node_contract = Self::get_node_contract(&contract.clone())?;
+        let mut deployment_contract = Self::get_deployment_contract(&contract.clone())?;
+        let cr_contract = Contracts::<T>::get(deployment_contract.capacity_reservation_id)
+            .ok_or(Error::<T>::ContractNotExists)?;
+        let capacity_reservation_contract = Self::get_capacity_reservation_contract(&cr_contract)?;
 
         // remove and reinsert contract id by node id and hash because that hash can have changed
         ContractIDByNodeIDAndHash::<T>::remove(
-            node_contract.node_id,
-            node_contract.deployment_hash,
+            capacity_reservation_contract.node_id,
+            deployment_contract.deployment_hash,
         );
         ContractIDByNodeIDAndHash::<T>::insert(
-            node_contract.node_id,
+            capacity_reservation_contract.node_id,
             &deployment_hash,
             contract_id,
         );
 
-        node_contract.deployment_hash = deployment_hash;
-        node_contract.deployment_data = deployment_data;
+        deployment_contract.deployment_hash = deployment_hash;
+        deployment_contract.deployment_data = deployment_data;
+
+        // update the resources with the extra resources
+        if let Some(resources) = resources {
+            if deployment_contract.resources != resources {
+                // first unclaim all resources from deployment contract
+                Self::_unclaim_resources_on_capacity_reservation(
+                    cr_contract.contract_id,
+                    deployment_contract.resources,
+                )?;
+                // then claim the new required resources
+                Self::_claim_resources_on_capacity_reservation(cr_contract.contract_id, resources)?;
+                deployment_contract.resources = resources;
+            }
+        }
 
         // override values
-        contract.contract_type = types::ContractData::NodeContract(node_contract);
-
-        let state = contract.state.clone();
-        Self::_update_contract_state(&mut contract, &state)?;
+        contract.contract_type = types::ContractData::DeploymentContract(deployment_contract);
+        contract.state = contract.state.clone();
+        Contracts::<T>::insert(&contract.contract_id, contract.clone());
 
         Self::deposit_event(Event::ContractUpdated(contract));
 
         Ok(().into())
     }
 
+    #[transactional]
```

Comment:
Why has this become a transactional?

Reply:
There is no more reason for it to be transactional. My bad.